src/main/java/com/mathworks/ci/RunMatlabTestsBuilder.java

```diff
@@ -0,0 +1,321 @@
+                                .warning(Message.getValue("Builder.matlab.modelcoverage.support.warning"));
+                    }
+                } catch (MatlabVersionNotFoundException e) {
+                    return FormValidation.warning(Message.getValue("Builder.invalid.matlab.root.warning"));
+                }
+            }
+            
+            
+            return FormValidation.ok();
+        };
+        
+        public FormValidation doCheckStmResultsChkBx(@QueryParameter boolean stmResultsChkBx) {
+            List<Function<String, FormValidation>> listOfCheckMethods =
+                    new ArrayList<Function<String, FormValidation>>();
+            if (stmResultsChkBx) {
+                listOfCheckMethods.add(chkSTMResultsSupport);
+            }
+            return getFirstErrorOrWarning(listOfCheckMethods);
+        }
+        
+        Function<String, FormValidation> chkSTMResultsSupport = (String matlabRoot) -> {
+            FilePath matlabRootPath = new FilePath(new File(matlabRoot));
+            rel = new MatlabReleaseInfo(matlabRootPath);
+            final MatrixPatternResolver resolver = new MatrixPatternResolver(matlabRoot);
+            if(!resolver.hasVariablePattern()) {
+                try {
+                    if (rel.verLessThan(MatlabBuilderConstants.BASE_MATLAB_VERSION_EXPORTSTMRESULTS_SUPPORT)) {
+                        return FormValidation
+                                .warning(Message.getValue("Builder.matlab.exportstmresults.support.warning"));
+                    }
+                } catch (MatlabVersionNotFoundException e) {
+                    return FormValidation.warning(Message.getValue("Builder.invalid.matlab.root.warning"));
+                }
+            }
+            
+            
+            return FormValidation.ok();
+        };
+        
+        public FormValidation getFirstErrorOrWarning(
+                List<Function<String, FormValidation>> validations) {
+            if (validations == null || validations.isEmpty())
+                return FormValidation.ok();
+            try {
+                final String matlabRoot = Jenkins.getInstance()
+                        .getDescriptorByType(AddMatlabToPathDescriptor.class).getMatlabRootFolder();
+                for (Function<String, FormValidation> val : validations) {
+                    FormValidation validationResult = val.apply(matlabRoot);
+                    if (validationResult.kind.compareTo(Kind.ERROR) == 0
+                            || validationResult.kind.compareTo(Kind.WARNING) == 0) {
+                        return validationResult;
+                    }
+                }
+            }catch (Exception e) {
+                return FormValidation.warning(Message.getValue("Builder.invalid.matlab.root.warning"));
+            }
+           
+            return FormValidation.ok();
+        }
+    }
+
+    @Override
+    public void perform(@Nonnull Run<?, ?> build, @Nonnull FilePath workspace,

```

Comment:
Both perform methods have the exact same implementation. Is the duplication unavoidable?

Reply:
This is because our execMatlabCommand() does differ for both the Builders. and build status does depends on what is the result of ExecMatlabCommand() hence  only step setEnv()  becomes redundant. also perform steps are pretty much builder specific even if it is redundant sometimes.